src/Xamarin.Android.Build.Tasks/Utilities/XAAssemblyResolver.cs

```diff
@@ -0,0 +1,333 @@
+			GC.SuppressFinalize (this);
+		}
+	}
+
+	/// <summary>
+	/// Contains a collection of directories where framework assemblies can be found.  This collection **must not**
+	/// contain any directories which contain ABI-specific assemblies.  For those, use <see cref="AbiSearchDirectories"/>
+	/// </summary>
+	public ICollection<string> FrameworkSearchDirectories { get; } = new List<string> ();
+
+	/// <summary>
+	/// Contains a collection of directories where Xamarin.Android (via linker, for instance) has placed the ABI
+	/// specific assemblies.  Each ABI has its own set of directories to search.
+	/// <summary>
+	public IDictionary<AndroidTargetArch, ICollection<string>> AbiSearchDirectories { get; } = new Dictionary<AndroidTargetArch, ICollection<string>> ();
+
+	readonly List<MemoryMappedViewStream> viewStreams = new List<MemoryMappedViewStream> ();
+	bool disposed;
+	TaskLoggingHelper log;
+	bool loadDebugSymbols;
+	ReaderParameters readerParameters;
+	readonly Dictionary<string, CacheEntry> cache;
+
+	public XAAssemblyResolver (TaskLoggingHelper log, bool loadDebugSymbols, ReaderParameters? loadReaderParameters = null)
+	{
+		this.log = log;
+		this.loadDebugSymbols = loadDebugSymbols;
+		this.readerParameters = loadReaderParameters ?? new ReaderParameters();
+
+		cache = new Dictionary<string, CacheEntry> (StringComparer.OrdinalIgnoreCase);
+	}
+
+	public AssemblyDefinition Resolve (string fullName, ReaderParameters? parameters = null)
+	{
+		return Resolve (AssemblyNameReference.Parse (fullName), parameters);
+	}
+
+	public AssemblyDefinition Resolve (AssemblyNameReference name)
+	{
+		return Resolve (name, null);
+	}
+
+	public AssemblyDefinition Resolve (AssemblyNameReference name, ReaderParameters? parameters)
+	{
+		return Resolve (AndroidTargetArch.None, name, parameters);
+	}
+
+	public AssemblyDefinition Resolve (AndroidTargetArch arch, AssemblyNameReference name, ReaderParameters? parameters = null)
+	{
+		string shortName = name.Name;
+		if (cache.TryGetValue (shortName, out CacheEntry? entry)) {
+			return SelectAssembly (arch, name.FullName, entry, loading: false);
+		}
+
+		if (arch == AndroidTargetArch.None) {
+			return FindAndLoadFromDirectories (arch, FrameworkSearchDirectories, name, parameters);
+		}
+
+		if (!AbiSearchDirectories.TryGetValue (arch, out ICollection<string>? directories) || directories == null) {
+			throw CreateLoadException (name);
+		}
+
+		return FindAndLoadFromDirectories (arch, directories, name, parameters);
+	}
+
+	AssemblyDefinition FindAndLoadFromDirectories (AndroidTargetArch arch, ICollection<string> directories, AssemblyNameReference name, ReaderParameters? parameters)
+	{
+		string? assemblyFile;
+		AssemblyDefinition? candidate = null;
+		foreach (string dir in directories) {
+			if ((assemblyFile = SearchDirectory (name.Name, dir)) != null) {
+				AssemblyDefinition? loaded = Load (arch, assemblyFile);
+				if (Array.Equals (loaded?.Name.MetadataToken, name.MetadataToken)) {
```

Comment:
While `DirectoryAssemblyResolver` has this code, I'm not certain that we actually want this anymore?  We have this "bizarre" [internal bug](https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1819836) in which `System.Runtime.dll` v6.0 can't be resolved, and while investigating that I found that [dotnet/linker *doesn't* requires that `.MetadataToken` match](https://github.com/dotnet/linker/blob/ba65e934dcb8cfbc81a494e890927f9a5d31deac/src/linker/Linker/AssemblyResolver.cs#L71-L87).  There is thus a case that we shouldn't mimic `DirectoryAssemblyResolver` and should instead mimic `AssemblyResolver`.

Reply:
I kept most of the code around because it has served us well over the years, meaning it was fine. However, if you think this code isn't needed anymore, I can remove it, sure.